pkg/controller/windowsmachineconfig/nodeconfig/nodeconfig.go

```diff
@@ -221,34 +241,25 @@ func (nc *nodeConfig) getNode(externalIP string) (*v1.Node, error) {
 		return nil, fmt.Errorf("no nodes found")
 	}
 
-	// Find the node that has the given IP
 	for _, node := range nodes.Items {
-		for _, address := range node.Status.Addresses {
-			if address.Type == "ExternalIP" && address.Address == externalIP {
-				matchedNode = &node
-				break
-			}
+		if instanceID == getInstanceID(node.Spec.ProviderID) {
+			return &node, nil
 		}
-		if matchedNode != nil {
-			break
-		}
-	}
-	if matchedNode == nil {
-		return nil, fmt.Errorf("could not find node with IP: %s", externalIP)
 	}
-	return matchedNode, nil
+	return nil, errors.Errorf("unable to find node for instanceID %s", instanceID)
 }
 
 // waitForNodeAnnotation checks if the node object has the given annotation and waits for retry.Interval seconds and
 // returns an error if the annotation does not appear in that time frame.
 func (nc *nodeConfig) waitForNodeAnnotation(nodeName, annotation string) error {
 	var found bool
 	err := wait.Poll(retry.Interval, retry.Timeout, func() (bool, error) {
-		node, err := nc.k8sclientset.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+		var err error
+		nc.node, err = nc.k8sclientset.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
```

Comment:
You're updating the nc.node directly here, you should wait till you get the node object, imagine an intermediate state where `Get` returns nil and you've now nc.node is nil, if there is another method which tries to access the nodeObject in meanwhile it will be problem, so update the nc.node only when you're sure about it.